bb8/src/inner.rs

```diff
@@ -117,8 +121,27 @@ where
         };
 
         match timeout(self.inner.statics.connection_timeout, future).await {
-            Ok(result) => result,
-            _ => Err(RunError::TimedOut),
+            Ok(result) => {
+                if get_direct {
+                    self.inner
```

Comment:
This is a little verbose for my taste. Suggest defining an `enum GetKind { Direct, Indirect, TimedOut }` and a method `SharedPool::record(kind: GetKind)` that abstracts over the details here.

Also I think this is missing records for the direct path around line 109 and 113?

Reply:
> This is a little verbose for my taste. Suggest defining an enum GetKind { Direct, Indirect, TimedOut } and a method SharedPool::record(kind: GetKind) that abstracts over the details here.

Yeps I can do that but this verbosity will be in some way unavoidable, within the `record` method we will need to branch for increasing one of the atomics, true that we can just use a `match` operator that could make this verbosity more shinny.

> Also I think this is missing records for the direct path around line 109 and 113?

It should not unless Im mistaken, in the end we are waiting for either the future to resolve or the timeout, so any result from the future if the timeout is not triggered will end up either as a `get_direct` or as a `get_waited`. Or am I missing something here?

Reply:
> Yeps I can do that but this verbosity will be in some way unavoidable, within the `record` method we will need to branch for increasing one of the atomics, true that we can just use a `match` operator that could make this verbosity more shinny.

Sure -- but `get()` is already pretty complex so it would be nice to move some of the complexity elsewhere. And it should be possible to write `record()` in a way that it doesn't need to repeat the `SeqCst` 3 times.

> > Also I think this is missing records for the direct path around line 109 and 113?
> 
> It should not unless Im mistaken, in the end we are waiting for either the future to resolve or the timeout, so any result from the future if the timeout is not triggered will end up either as a `get_direct` or as a `get_waited`. Or am I missing something here?

No, you're right! Sorry, was thinking of an earlier version of this code.

Reply:
Oki doki, Ill make the changes later!